### PR TITLE
Allow async ops for all gather with gather dim != 0

### LINKED ISSUE #136425: Any updates on AsyncCollectiveTensor support for all-gather along non-zero dims?
https://github.com/pytorch/pytorch/blob/e9bfbf78d5d89df1ec59cb82d7f78b85f9014a98/torch/distributed/_functional_collectives.py#L208

### PR DESCRIPTION
**Pull request by mayank31398**
Fixes #136425
as title

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Comment by awgu** [comment]
cc: @yifuwang do you remember why `trigger_wait` vs. just calling `wait`?

**Comment by yifuwang** [comment]
@awgu not really. I think it's here because it was in the early prototype. We should probably get rid of it.

**Comment by yifuwang** [comment]
@bdhirsh do you know if we ever trace through ACT? I have the impression that we don't use ACT during tracing, but I'm not totally sure. Asking because this is something you might want to take a look if we trace through ACT.

**Comment by mayank31398** [comment]
btw, is it not possible to just avoid this?
is there a fundamental limitation of not being able to all gather along dim=1

**Comment by awgu** [comment]
All-gather is fundamentally a dim-0 operation 🤔

**Comment by yifuwang** [comment]
@awgu I think there's an optimization if we don't require contiguous stride output: we could perform `.movedim(0, dim).contiguous()` on the input shard before the all-gather, then perform `.as_strided()` on the output. This way each rank only performs layout transformation on a tensor `world_size`x smaller (plus it's a view-only epilogue).

**Comment by bdhirsh** [comment]
> do you know if we ever trace through ACT? I have the impression that we don't use ACT during tracing, but I'm not totally sure. Asking because this is something you might want to take a look if we trace through ACT.

@yifuwang so there are actually two cases:

(1) if we encounter a functional collective during tracing, then we will directly emit the `wait_tensor()` call, without ever constructing an ACT ([link](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_functional_collectives.py#L814))

(2) **however** there is a special case: if a collective ran in eager mode, and that collective output becomes an input to the compiled region, then we will see that one of our graph inputs is an `AsyncCollectiveTensor()`, and the first time it is used in the compiled function we will trace into its `__torch_dispatch__`, tracing out a `wait_tensor()` (that inductor could theoretically move to later in the compiled region, if it is profitable)

File: torch/distributed/_functional_collectives.py [hunk 1]
```diff
@@ -176,6 +177,10 @@ def all_reduce(self: torch.Tensor, reduceOp: str, group: RANK_TYPES, tag: str =
     return _maybe_wrap_tensor(tensor)
 
 
+def callback(x, group_size, gather_dim):
+    return torch.cat(torch.chunk(x, group_size, dim=0), dim=gather_dim)
```
Comment (bdhirsh) [inline torch/distributed/_functional_collectives.py:181]: Hmm... is the idea that we want to delay the `cat() and `chunk()` calls until the result of the allgather is actually needed?

We might want to be a bit careful about the (2) compile case I mentioned in the other comment. In particular, could you write a test like this?
```
@torch.compile
def g(x):
    return torch.matmul(x, x)
    
def f(x):
    y = allgather(x)
    return g(y, y)
```

where if you run it with `TORCH_LOGS="aot"`, I think what we'd want is for the `wait_tensor()` + `cat()` + `chunk()` to all be invoked lazily, so they actually show up inside of the graph that compile generates (before the matmul).

**Comment by mayank31398** [comment]
btw, we should also run the callback on a separate stream
its currently blocking the computation

**Comment by awgu** [comment]
I think overlapping the callback by running it in a separate stream is tricky memory-wise. Overlapping it with compute might not be better especially if the next compute op is a matmul.

**Comment by mayank31398** [comment]
@awgu can you clarify the point about the matmul?
is it because overlapping the collective with matmul slows matmul down?
I have seen such behaviour and I am not sure of the reason

**Comment by awgu** [comment]
Yes. I think because there might be some memory contention between the device copy and the subsequent matmul.

File: torch/distributed/_functional_collectives.py (hunk 1 above)
Comment (kwen2501) [reply torch/distributed/_functional_collectives.py:181]: nit: can you add comment to this function? Thanks.

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -601,29 +610,49 @@ def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
...
             out = wait_tensor(self.elem)
             self.completed = True
-            return out
         else:
-            return self.elem
+            out = self.elem
 
-    def wait(self) -> torch.Tensor:
-        return wait_tensor(self.elem)
+        return self._run_callbacks(out)
```
Comment (kwen2501) [inline torch/distributed/_functional_collectives.py:622]: Do the new `wait()` and the old `wait()` have the same semantics?

File: torch/distributed/_functional_collectives.py [hunk 3]
```diff
@@ -601,29 +610,49 @@ def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
...
 
     def _get_acs_underlying_tensor(self):
         """This method enables  _functional_collectives_impl to test if a tensor is an ACS"""
         return self.elem
 
     @classmethod
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        args_tuple = []
+        for arg in args:
+            if isinstance(arg, AsyncCollectiveTensor) and len(arg.callbacks) > 0:
```
Comment (kwen2501) [inline torch/distributed/_functional_collectives.py:639]: what if `len(arg.callbacks) == 0`, then we don't wait for ACT?

Comment (mayank31398) [reply torch/distributed/_functional_collectives.py:657]: hey @kwen2501 I am not familiar with what you mean by ACT, can you explain?

Comment (mayank31398) [reply torch/distributed/_functional_collectives.py:657]: yeah, right now I think we dont need to but there is some incorrectness happening I think
Ill look into it
